src/VisualStudio/Core/Def/Implementation/ProjectSystem/Readme.md

```diff
@@ -0,0 +1,41 @@
+﻿# Locking and Synchronization
+
+There are two core types in this folder, each of which has their own lock.
+
+Each project from the project system is represented by a VisualStudioProject which has as SemaphoreSlim _gate. This lock is taken
+any time a mutation happens to that individual project, to ensure that the type itself is safe to use concurrently. The expectation
+though is that simultaneous use from multiple threads by a projet system isn't common, so no real effort has been expended to try
+to make the locking fine-grained there; each function just acquires the _gate and then does what it needs to do.
+
+The workspace (VisualStudioWorkspaceImpl) also has it's own SemaphoreSlim _gate. This is acquired any time a change is being made
+to the workspace. As much as possible, try to acquire this gate in an asynchronous fashion, since during solution load this lock
+will have a lot of things trying to acquire it at once on a bunch of threads, and we may starve off the thread pool if we're not
+careful. Unfortunately however, we still have legacy project systems that aren't async friendly; they still may apply changes or
+batches synchronously so in those cases we still acquire the gate in a synchronous fashion.
```

Comment:
why do we still have these legacy project systems?

Reply:
CPS doesn't support various project types these days.

Reply:
if they dont' support them, do we need to?  basically, this all seems to go against all the feedback we're getting that things need to be fast, or else we can't compete.

Reply:
It's been over a year since I saw the telemetry, but legacy projects are probably still the majority of projects opened in the wild. Presumably not _new_ projects created though.